jetty-io/src/main/java/org/eclipse/jetty/io/CyclicTimeouts.java

```diff
@@ -0,0 +1,147 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.eclipse.jetty.util.thread.Scheduler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A {@link CyclicTimeout} that manages the timeouts for many entities.</p>
+ * <p>A typical scenario is for a parent entity to manage the timeouts of many children entities.</p>
+ * <p>When a new entity is created, call {@link #schedule(T)} with the new entity so that
+ * this instance can be aware and manage the timeout of the new entity.</p>
+ * <p>Eventually, this instance wakes up and iterates over the entities provided by {@link #iterator()}.
+ * During the iteration, each entity:</p>
+ * <ul>
+ *   <li>may never expire (see {@link Expirable#getExpireNanoTime()}; the entity is ignored</li>
+ *   <li>may be expired; {@link #onExpired(T)} is called with that entity as parameter</li>
+ *   <li>may expire at a future time; the iteration records the earliest expiration time among
+ *   all non-expired entities</li>
+ * </ul>
+ * <p>When the iteration is complete, this instance is re-scheduled with the earliest expiration time
+ * calculated during the iteration.</p>
+ *
+ * @param <T> the entity type
+ */
+public abstract class CyclicTimeouts<T extends CyclicTimeouts.Expirable> extends CyclicTimeout implements Iterable<T>
+{
```

Comment:
I'm not sure this class is-a CyclicTimeout.  For a start this class works on expiry times rather than delays.  Also we don't want users calling the underlying schedule method directly.
```suggestion
public abstract class CyclicTimeouts<T extends CyclicTimeouts.Expirable> implements Iterable<T>
{
    private final CyclicTimeout timout;
```

Reply:
Done.
Also remove is-a `Iterable` for the same reasons, only subclasses should override `iterator()` and only the implementation should call it.